ibverbs/src/lib.rs

```diff
@@ -368,33 +372,53 @@ impl Context {
     /// A protection domain is a means of protection, and helps you create a group of object that
     /// can work together. If several objects were created using PD1, and others were created using
     /// PD2, working with objects from group1 together with objects from group2 will not work.
-    pub fn alloc_pd(&self) -> Result<ProtectionDomain, ()> {
+    pub fn alloc_pd(self: Arc<Self>) -> Result<ProtectionDomain, ()> {
         let pd = unsafe { ffi::ibv_alloc_pd(self.ctx) };
         if pd.is_null() {
             Err(())
         } else {
-            Ok(ProtectionDomain { ctx: self, pd })
+            Ok(ProtectionDomain {
+                ctx: self.clone(),
+                pd,
+            })
+        }
+    }
+
+    /// Query corresponding device
+    pub fn query_device(&self) -> io::Result<DeviceAttr> {
+        let mut dev_attrs = Default::default();
+
+        let ret = unsafe { ffi::ibv_query_device(self.ctx, &mut dev_attrs) };
+
+        if ret != 0 {
+            return Err(io::Error::last_os_error());
         }
+
+        Ok(dev_attrs)
     }
 }
 
 impl Drop for Context {
     fn drop(&mut self) {
-        let ok = unsafe { ffi::ibv_close_device(self.ctx) };
+        let ok = if !self.ctx.is_null() {
```

Comment:
How can `ctx` even be NULL here? Wouldn't that mean the context failed to be constructed in the first place?

Reply:
This one I'm not 100% sure on. Let me see if I can find the original commit and whether that explains it